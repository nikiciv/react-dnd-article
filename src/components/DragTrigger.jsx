/* @flow */
import React from 'react';

// * isBodyTrigger: use 'body' as a trigger or if not specified the component itself
// * children: render props function
// * className: offers possibility to style the component according to the context
// * style: offers possibility to add styles
type Props = {
  isBodyTrigger: boolean,
  children: func,
  className: string,
  style: object,
};

type State = {
  isDragActive: boolean,
};

// This component will notify when the user is dragging a file over the trigger element
class DragTrigger extends React.PureComponent<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      isDragActive: false,
    };
    this.counter = 0;
    this.isBodyTrigger = this.props.isBodyTrigger;
  }

  componentDidMount = () => {
    if (this.isBodyTrigger) {
      document.addEventListener('dragenter', this.handleDragEnter);
      document.addEventListener('dragleave', this.handleDragLeave);
      document.addEventListener('drop', this.handleDrop);
    }
  };

  componentWillUnmount = () => {
    document.removeEventListener('dragenter', this.handleDragEnter);
    document.removeEventListener('dragleave', this.handleDragLeave);
    document.removeEventListener('drop', this.handleDrop);
  };

  handleDragEnter = () => {
    this.counter++;
    this.setState({ isDragActive: true });
  };

  handleDragLeave = () => {
    this.counter--;
    if (this.counter === 0) {
      this.setState({ isDragActive: false });
    }
  };

  handleDrop = () => {
    setTimeout(() => {
      this.setState({ isDragActive: false });
    }, 0);
    this.counter = 0;
  };

  render() {
    const internalProps = {};

    if (!this.isBodyTrigger) {
      internalProps.onDragEnter = this.handleDragEnter;
      internalProps.onDragLeave = this.handleDragLeave;
      internalProps.onDrop = this.handleDrop;
    }

    return (
      <div className={this.props.className} style={this.props.style} {...internalProps}>
        {this.props.children(this.state.isDragActive)}
      </div>
    );
  }
}

export default DragTrigger;
