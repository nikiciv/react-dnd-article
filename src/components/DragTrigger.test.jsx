import React from 'react';
import { mount } from 'enzyme';
import DragTrigger from './DragTrigger';

describe('<DragTrigger />', () => {
  let props;
  let instance;

  props = {
    isBodyTrigger: true,
    children: jest.fn(),
  };

  beforeEach(() => {
    instance = mount(<DragTrigger {...props} />);
  });

  it('renders DragTrigger component correctly', () => {
    expect(instance).toMatchSnapshot();
  });

  it('sets isDragActive to false when an element is dragged out of the browser', () => {
    expect(instance.state().isDragActive).toBe(false);
    const dragEnterEvent = new Event('dragenter');
    document.dispatchEvent(dragEnterEvent);
    expect(instance.state().isDragActive).toBe(true);
    const dragLeaveEvent = new Event('dragleave');
    document.dispatchEvent(dragLeaveEvent);
    expect(instance.state().isDragActive).toBe(false);
  });

  it('triggers isDragActive when an element is dragged in the drag area', () => {
    props = {
      isBodyTrigger: false,
      children: jest.fn(),
    };
    instance = mount(<DragTrigger {...props} />);
    const dragEnterEvent = new Event('dragenter');
    document.dispatchEvent(dragEnterEvent);
    expect(instance.state().isDragActive).toBe(false);
    instance.simulate('dragenter');
    expect(instance.state().isDragActive).toBe(true);
  });

  it('sets isDragActive to false when an element is dragged out of the drag area', () => {
    instance.setProps({
      isBodyTrigger: false,
    });
    expect(instance.props().isBodyTrigger).toBe(false);
    instance.simulate('dragenter');
    expect(instance.state().isDragActive).toBe(true);
    instance.simulate('dragleave');
    expect(instance.state().isDragActive).toBe(false);
  });
});
