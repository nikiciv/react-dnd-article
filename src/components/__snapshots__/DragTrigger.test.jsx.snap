// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DragTrigger /> renders DragTrigger component correctly 1`] = `
<DragTrigger
  isBodyTrigger={true}
>
  <div />
</DragTrigger>
`;
